antlr4cpp/antlr/v4/runtime/atn/atn_deserializer.cpp

```diff
@@ -0,0 +1,1335 @@
+				return std::make_shared<rule_transition>(std::static_pointer_cast<rule_start_state>(atn->states()[static_cast<size_t>(arg1)]), arg2, arg3, target);
+
+			case transition::transition_type::predicate:
+				return std::make_shared<predicate_transition>(target, arg1, arg2, !!arg3);
+
+			case transition::transition_type::precedence:
+				return std::make_shared<precedence_transition>(target, arg1);
+
+			case transition::transition_type::atom:
+				return std::make_shared<atom_transition>(target, arg3 ? token::eof : arg1);
+
+			case transition::transition_type::action:
+				return std::make_shared<action_transition>(target, arg1, arg2, !!arg3);
+
+			case transition::transition_type::set:
+				return std::make_shared<set_transition>(target, sets[static_cast<size_t>(arg1)]);
+
+			case transition::transition_type::not_set:
+				return std::make_shared<not_set_transition>(target, sets[static_cast<size_t>(arg1)]);
+
+			case transition::transition_type::wildcard:
+				return std::make_shared<wildcard_transition>(target);
+
+			default:
+				assert(!"Invalid transition type.");
+				return nullptr;
+			}
+		}
+
+		std::shared_ptr<lexer_action> lexer_action_factory(lexer_action::lexer_action_type type, int32_t data1, int32_t data2)
+		{
+			switch (type)
+			{
+			case lexer_action::lexer_action_type::channel:
+				return std::make_shared<lexer_action::channel_action>(data1);
+
+			case lexer_action::lexer_action_type::custom:
+				return std::make_shared<lexer_action::custom_action>(data1, data2);
+
+			case lexer_action::lexer_action_type::mode:
+				return std::make_shared<lexer_action::mode_action>(data1);
+
+			case lexer_action::lexer_action_type::more:
+				return lexer_action::more_action::instance;
+
+			case lexer_action::lexer_action_type::pop_mode:
+				return lexer_action::pop_mode_action::instance;
+
+			case lexer_action::lexer_action_type::push_mode:
+				return std::make_shared<lexer_action::push_mode_action>(data1);
+
+			case lexer_action::lexer_action_type::skip:
+				return lexer_action::skip_action::instance;
+
+			case lexer_action::lexer_action_type::type:
+				return std::make_shared<lexer_action::type_action>(data1);
+
+			case lexer_action::lexer_action_type::indexed_custom:
+			default:
+				assert(!"Invalid action type.");
+				return nullptr;
+			}
+		}
+
+	}
+
+	std::shared_ptr<grammar_atn> atn_deserializer::deserialize(std::vector<uint16_t> const& raw_data) const
+	{
+		std::vector<uint16_t> data(raw_data);
+
+		// don't adjust the first value since that's the version number
+		for (size_t i = 1; i < data.size(); i++)
+		{
+			data[i] = static_cast<uint16_t>(data[i] - 2);
+		}
+
+		size_t p = 0;
+		int32_t version = to_int(data[p++]);
+		if (version != serialized_version)
+		{
+			//String reason = String.format(Locale.getDefault(), "Could not deserialize ATN with version %d (expected %d).", version, SERIALIZED_VERSION);
+			throw std::runtime_error("error message not implemented");
```

Comment:
Nasty old vsnprintf() comes to mind; or use a std::strstream, which is a bit more modern.